app/src/app-logic.js

```diff
@@ -28,34 +30,47 @@ export default function useAppLogic() {
       convictionStakes.filter(({ entity }) => entity === connectedAccount)) ||
     []
 
-  const myCurrentStakes = useMemo(() => {
-    return myStakesHistory.reduce(
-      (stakes, { proposal: currProposalId, tokensStaked }) => {
-        if (tokensStaked === 0) stakes.delete(currProposalId)
-        else {
-          const proposal = proposals.find(({ id }) => id === currProposalId)
-          if (proposal && !proposal.executed) {
-            stakes.set(
-              currProposalId,
-              formatTokenAmount(
-                parseInt(tokensStaked),
-                parseInt(stakeToken.tokenDecimals)
-              )
-            )
-          }
-        }
-        return stakes
-      },
-      new Map()
-    )
-  }, [myStakesHistory.length])
+  const myActiveStakes = useMemo(() => {
+    if (!connectedAccount || !stakeToken.tokenDecimals || !proposals) {
+      return
+    }
+    return proposals.reduce((myStakes, proposal) => {
+      if (proposal.executed || !proposal.stakes) {
+        return myStakes
+      }
+
+      const totalActive = proposal.stakes.reduce((accumulator, stake) => {
+        return accumulator.add(stake.amount)
+      }, new BN('0'))
+
+      setTotalActiveTokens(prevValue => prevValue.add(totalActive))
+
+      const myStake = proposal.stakes.find(
+        stake => stake.entity === connectedAccount
+      )
+      if (!myStake) {
+        return myStakes
+      }
+
+      myStakes.push({
+        proposal: proposal.id,
+        proposalName: proposal.name,
+        stakedAmount: myStake.amount,
+      })
+      setMyActiveTokens(prevValue => prevValue.add(myStake.amount))
+      return myStakes
+    }, [])
+  }, [proposals.length, connectedAccount, stakeToken.tokenDecimals])
```

Comment:
Thinking about this bit of code, looks a little bit weird to set states on a `useMemo`. 
Use memos are meant for optimization when computing values, and probably we shouldn't do any side effects inside it.

More over looks like we can just return the `totalActive` as part of the computed values along with `myActiveStakes` and not have another state for something that is computed from other state variable.

The other thing that maybe we could lift down to the `StakingTokens` component is the `myActiveTokens` variable, since this one could be computed directly from `myStakes` and i think is only used there? This way we avoid another variable passed down as props.

Also, i think that the whole `proposals` variable should be the `useMemo` dependency as opposed to the `proposals.length` since for example when a staking is done, the computed value will not be re-calculated.

Reply:
mhh not sure about that, why we need to re run a reduce to do that math on the component render time? if we can do it here and all in once? do you have some reference where i can read about what you are describing about the `useMemo`? trying to search and couldn't find anything